### PR TITLE
Rewrite `AutosuggestTextarea` as Functional Component

### PR DESCRIPTION
**Pull request by ClearlyClaire**
Rewrite `AutosuggestTextarea` as a Functional Component and move a lot of `state` to refs.

This fixes the Input Method issues introduced by #25047 in a way similar to #27613 by avoiding the `setState` flow, although it does not address the underlying issue of the `children` prop changing in-between two events for no discernable reason.

We were using `ImmutablePureComponent` but only benefitted from it for the `suggestions` prop, which in practice should not be spuriously recomputed, so moving away from it should be ok.

Something a bit awkward is the use of `useEffect`, this is to reproduce the functionality brought by `UNSAFE_componentWillReceiveProps` (resetting `suggestionsHidden` when `suggestions` change provided that the textarea is focused).

Another notable change is the use of `forwardRef` because the parent component uses a reference to the textarea.

### CODE REVIEW

File: app/javascript/mastodon/components/autosuggest_textarea.jsx [hunk 1]
```diff
@@ -102,80 +96,78 @@ export default class AutosuggestTextarea extends ImmutablePureComponent {
...
-    const suggestion = this.props.suggestions.get(e.currentTarget.getAttribute('data-index'));
+  const handleSuggestionClick = useCallback((e) => {
+    const suggestion = suggestions.get(e.currentTarget.getAttribute('data-index'));
     e.preventDefault();
-    this.props.onSuggestionSelected(this.state.tokenStart, this.state.lastToken, suggestion);
-    this.textarea.focus();
-  };
+    onSuggestionSelected(tokenStartRef.current, lastTokenRef.current, suggestion);
+    textareaRef.current?.focus();
+  }, [suggestions, onSuggestionSelected, textareaRef]);
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/components/autosuggest_textarea.jsx:153]: This doesn't work, `textareaRef`, as passed by the upper component, is a function, not a ref…

**Review by renchap — DISMISSED** [review]
Looks good to me, and much cleaner

**Comment by zunda** [comment]
This, with reverting other fixes, worked for me. Thank you!

**Comment by ClearlyClaire** [comment]
Removed from the queue as I just noticed a regression. Going to fix that real quick.

**Comment by ClearlyClaire** [comment]
> Removed from the queue as I just noticed a regression. Going to fix that real quick.

Addressed

**Review by renchap — APPROVED** [review]

**Comment by mgmn** [comment]
Oops, this looks prettier, but I deployed it disallows me to pick an emoji from the picker and this not only in `/publish`...

File: app/javascript/mastodon/features/compose/components/compose_form.jsx [hunk 2]
```diff
@@ -214,7 +216,7 @@ class ComposeForm extends ImmutablePureComponent {
 
   handleEmojiPick = (data) => {
     const { text }     = this.props;
-    const position     = this.autosuggestTextarea.textarea.selectionStart;
+    const position     = this.textarea.selectionStart;
```
Comment (mgmn) [inline app/javascript/mastodon/features/compose/components/compose_form.jsx:219]: I think we need to just replace `textarea` with `textareaRef.current`

Comment (mgmn) [reply app/javascript/mastodon/features/compose/components/compose_form.jsx:219]: I've created #27647 where gets this done.

**Pull request merged**